Source/Core/VideoCommon/VertexLoaderX64.cpp

```diff
@@ -0,0 +1,453 @@
+			load_bytes = 2;
+			break;
+
+		case FORMAT_16B_4444:
+			//                   RRRRGGGG BBBBAAAA
+			// AAAAAAAA BBBBBBBB GGGGGGGG RRRRRRRR
+			LoadAndSwap(16, scratch1, data);
+			if (cpu_info.bBMI2)
+			{
+				PDEP(32, scratch2, scratch1, M(&mask_0f));
+				PDEP(32, scratch1, scratch1, M(&mask_f0));
+			}
+			else
+			{
+				MOV(32, R(scratch3), R(scratch1));
+				SHL(32, R(scratch1), Imm8(12));
+				AND(32, R(scratch1), Imm32(0x0F000000));
+				MOV(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
+				SHL(32, R(scratch1), Imm8(8));
+				AND(32, R(scratch1), Imm32(0x000F0000));
+				OR(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
+				SHL(32, R(scratch1), Imm8(4));
+				AND(32, R(scratch1), Imm32(0x00000F00));
+				OR(32, R(scratch2), R(scratch1));
+
+				AND(32, R(scratch3), Imm8(0x0F));
+				OR(32, R(scratch2), R(scratch3));
+
+				MOV(32, R(scratch1), R(scratch2));
+				SHL(32, R(scratch1), Imm8(4));
+			}
+			OR(32, R(scratch1), R(scratch2));
+			SwapAndStore(32, MDisp(dst_reg, m_dst_ofs), scratch1);
+			load_bytes = 2;
+			break;
+
+		case FORMAT_24B_6666:
+			//          RRRRRRGG GGGGBBBB BBAAAAAA
+			// AAAAAAAA BBBBBBBB GGGGGGGG RRRRRRRR
+			data.offset -= 1;
+			LoadAndSwap(32, scratch1, data);
+			if (cpu_info.bBMI2)
+			{
+				PDEP(32, scratch1, scratch1, M(&mask_fc));
+				MOV(32, R(scratch2), R(scratch1));
+			}
+			else
+			{
+				MOV(32, R(scratch3), R(scratch1));
+				SHL(32, R(scratch1), Imm8(8));
+				AND(32, R(scratch1), Imm32(0xFC000000));
+				MOV(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
+				SHL(32, R(scratch1), Imm8(6));
+				AND(32, R(scratch1), Imm32(0x00FC0000));
+				OR(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
+				SHL(32, R(scratch1), Imm8(4));
+				AND(32, R(scratch1), Imm32(0x0000FC00));
+				OR(32, R(scratch2), R(scratch1));
+
+				SHL(32, R(scratch3), Imm8(2));
+				AND(32, R(scratch3), Imm32(0x000000FC));
+				OR(32, R(scratch2), R(scratch3));
+
+				MOV(32, R(scratch1), R(scratch2));
+			}
+
+			SHR(32, R(scratch1), Imm8(6));
+			AND(32, R(scratch1), Imm32(0x03030303));
+			OR(32, R(scratch1), R(scratch2));
+
+			SwapAndStore(32, MDisp(dst_reg, m_dst_ofs), scratch1);
+			load_bytes = 3;
+			break;
+	}
+	if (attribute == DIRECT)
+		m_src_ofs += load_bytes;
+}
+
+void VertexLoaderX64::GenerateVertexLoader()
+{
+	m_native_components = 0;
+	memset(&m_native_vtx_decl, 0, sizeof(m_native_vtx_decl));
+
+	// Backup count since we're going to count it down.
+	PUSH(32, R(ABI_PARAM3));
+
+	// We use ABI_PARAM3 for scratch2.
+	MOV(32, R(count_reg), R(ABI_PARAM3));
+
+	if (m_VtxDesc.Position & MASK_INDEXED)
+		XOR(32, R(skipped_reg), R(skipped_reg));
+
+	// TODO: load constants into registers outside the main loop
+
+	const u8* loop_start = GetCodePtr();
+
+	u32 posmatidx_ofs = 0;
+	if (m_VtxDesc.PosMatIdx)
+		posmatidx_ofs = m_src_ofs++;
+
+	u32 texmatidx_ofs[8];
+	const u64 tm[8] = {
```

Comment:
Could this be a BitSet8? Do we need tm[i] and friends for any purpose other than boolean if statements? It might be cleaner/simpler to do it that way.